libs/spec/Elf.3d

```diff
@@ -0,0 +1,460 @@
+
+  These two tables then describe the rest of the file. Roughly:
+              -------------------------------
+             |                               |
+             |                  -------------|------------------
+             |                 |             |                  |
+             |                 |            \ /                \ /
+   ----------------------------------------------------------------------
+  |..|ph_offset|ph_num|..|sh_offset|sh_num|..|ph table|file data|sh table|
+   ----------------------------------------------------------------------
+
+  |<---------------ELF Header--------------->|
+
+  Where ph_table contains ph_num entries, sh_table contains sh_num entries,
+  and both describe sections in the rest of the file (i.e. in the file data part)
+
+  This specification validates the:
+  - ELF header
+  - Program Header table (that it contains ph_num entries,
+                          and points to valid parts of the rest of the file)
+  - Section Header table (that it contains sh_num entries,
+                          and points to valid parts of the rest of the file)
+
+  Though it is possible to do it in 3d, we do not validate individual sections/segments yet.
+
+  See the ELF type at the end, that's the entrypoint for the specification
+*/
+
+
+#define MAX_UINT32 0xffffffff
+
+typedef UINT8 UCHAR;
+
+typedef struct _ZeroByte
+{
+    UINT8 zero { zero == 0 };
+} ZeroByte;
+
+// Address type is either UINT64 or UINT32 depending on the architecture
+// For now, the spec supports only 64 bit arch, later we could parameterize
+//   the Address type
+
+// The same goes for the OFFSET type
+
+typedef UINT64 ADDRESS;
+
+typedef UINT64 OFFSET;
+
+//
+// The first 4 bytes of the e_ident array in the ELF header
+//
+
+#define ELFMAG0    0x7f
+#define ELFMAG1    0x45  //'E'
+#define ELFMAG2    0x4c  //'L'
+#define ELFMAG3    0x46  //'F'
+
+//
+// The largest allowed value of the E_PHNUM field in the ELF Header
+//
+
+#define PN_XNUM        0xffff
+
+//
+// The largest allowed value of the E_SHNUM field in the ELF Haeder
+//
+
+#define SHN_LORESERVE  0xff00
+
+//
+// Value of E_SHSTRNDX when the ELF file has no Section Header table
+//
+
+#define SHN_UNDEF      0
+
+//
+// Enums for ELF class, data, OS ABI
+// Used to type the fields in the ELF header
+//
+
+UINT8 enum ELFCLASS
+{
+  ELFCLASSNONE = 0,
+  ELFCLASS32,
+  ELFCLASS64
+}
+
+UINT8 enum ELFDATA
+{
+  ELFDATANONE = 0,
+  ELFDATA2LSB,
+  ELFDATA2MSB
+}
+
+UINT8 enum ELFOSABI
+{
+  ELFOSABI_NONE = 0,
+  ELFOSABI_SYSV,
+  ELFOSABI_HPUX,
+  ELFOSABI_NETBSD,
+  ELFOSABI_LINUX,
+  ELFOSABI_SOLARIS,
+  ELFOSABI_IRIX,
+  ELFOSABI_FREEBSD,
+  ELFOSABI_TRU64,
+  ELFOSABI_ARM,
+  ELFOSABI_STANDALONE,
+}
+
+//
+// Last 7 bytes of the ELF header are 0s
+//
+
+#define E_IDENT_PADDING_SIZE    7
+
+
+/*++
+  The e_ident array in the ELF header
+  --*/
+
+typedef struct _E_IDENT
+{
+  //The first four bytes are 0x7f followed by 'E', 'L', and 'F'
```

Comment:
nit: comments are inconsistent in terms of whether a space appears after the // throughout this file